### PR TITLE
[HOTFIX] Fixed NPE during query with Local Dictionary

### PR DESCRIPTION
**Pull request by kumarvishal09**
**Problem:**
Query is failing with NPE when some blocklet encoded with local dictionary and some without local dictionary.
**Root Cause:** 
This is coming because in carbonvectorProxy setDictionary with null it is not setting the dictionary to null because of this it is treated like a local dictionary column but column is not encoded with dictionary.
**Solution:**
Set dictionary to null
 - [ ] Any interfaces changed?
 
 - [ ] Any backward compatibility impacted?
 
 - [ ] Document update required?

 - [ ] Testing done
        Please provide details on 
        - Whether new unit test cases have been added or why no new tests are required?
        - How it is tested? Please attach test report.
        - Is it a performance related change? Please attach the performance test report.
        - Any additional information to help reviewers in testing this change.
       
 - [ ] For large changes, please consider breaking it into sub-tasks under an umbrella JIRA.

### CODE REVIEW

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder2.1/2269/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://95.216.28.178:8080/job/ApacheCarbonPRBuilder1/2488/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.3.2, Please check CI http://136.243.101.176:8080/job/carbondataprbuilder2.3/10527/

**Comment by qiuchenjian** [comment]
why does one segment have some blocklet encoded with local dictionary and some without local dictionary ?

**Comment by ravipesala** [comment]
> why does one segment have some blocklet encoded with local dictionary and some without local dictionary ?

It is because carbon generates a dictionary based on the column value count threshold, so once it reaches that threshold it stops generating the dictionary. There are scenarios where some blocks/blocklets are with in threshold and some are not, thats why some blocks has local dictionary and some don't have

**Comment by ravipesala** [comment]
LGTM

**Comment by qiuchenjian** [comment]
@ravipesala I remember if the local dictionary value count reach threshold, it will go back to original value， right？
Does some nodes's local dictionary reach the threshold, bug other's not, so this scene appeared ？（some shards have local dictionary, some shards don't have  local dictionary）

File: integration/spark-datasource/src/main/spark2.1andspark2.2/org/apache/spark/sql/CarbonVectorProxy.java [hunk 1]
```diff
@@ -311,6 +311,7 @@ public void setDictionary(CarbonDictionary dictionary) {
         vector.setDictionary(dictionaryWrapper);
         this.dictionary = dictionaryWrapper;
       } else {
+        this.dictionary = null;
```
Comment (xuchuanyin) [inline integration/spark-datasource/src/main/spark2.1andspark2.2/org/apache/spark/sql/CarbonVectorProxy.java]: This change lies in spark2.1andspark2.2, so spark2.3 does not have this problem？

**Comment by ravipesala** [comment]
> @ravipesala I remember if the local dictionary value count reach threshold, it will fallback to original value， right？
> what level does it fallback? blocklet, page or block ?

The fallback is at page level.

**Comment by qiuchenjian** [comment]
> > @ravipesala I remember if the local dictionary value count reach threshold, it will fallback to original value， right？
> > what level does it fallback? blocklet, page or block ?
> 
> The fallback is at page level.

get it, thanks

File: integration/spark-datasource/src/main/spark2.1andspark2.2/org/apache/spark/sql/CarbonVectorProxy.java (hunk 1 above)
Comment (kumarvishal09) [reply integration/spark-datasource/src/main/spark2.1andspark2.2/org/apache/spark/sql/CarbonVectorProxy.java]: In 2.3 proper interface are exposed from spark and same has been implemented, so its not required

**Comment by ravipesala** [comment]
LGTM

**Pull request closed without merging**